File: learnsynth_offline_llm/android/src/main/kotlin/com/learnsynth/learnsynth_offline_llm/ModelManager.kt
package com.learnsynth.learnsynth_offline_llm

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.security.DigestInputStream
import java.security.MessageDigest


data class ManifestFile(
    val name:String,
    val size:Long,
    val sha256:String,
    val url:String?,
    val asset:Boolean,
    val demo:Boolean
)
data class Manifest(val ctxLen:Int, val tokenizer:String, val files:List<ManifestFile>)

class ModelManager(private val context: Context, private val modelId: String) {
    private val TAG = "ModelManager"
    private val rootDir = File(context.filesDir, "models/$modelId")

    fun modelDir(): File = rootDir

    suspend fun ensureModel(manifest: Manifest, progress: (Map<String, Any?>) -> Unit) = withContext(Dispatchers.IO) {
        if (!rootDir.exists()) rootDir.mkdirs()
        for (f in manifest.files) {
            val out = File(rootDir, f.name)
            if (f.demo) {
                out.parentFile?.mkdirs()
                FileOutputStream(out).use { }
                progress(mapOf("stage" to "demo-generate", "file" to f.name, "received" to f.size, "total" to f.size, "percent" to 100))
            } else if (f.asset) {
                progress(mapOf("stage" to "copy-asset", "file" to f.name, "received" to 0, "total" to f.size, "percent" to 0))
                copyAsset("assets/models/$modelId/${f.name}", out)
            } else {
                if (!out.exists() || !verifySha256(out, f.sha256)) {
                    if (f.url.isNullOrBlank()) throw IOException("Missing URL for ${f.name}")
                    downloadTo(f.url!!, out, f.size) { rec, tot, pct ->
                        progress(mapOf("stage" to "downloading", "file" to f.name, "received" to rec, "total" to tot, "percent" to pct))
                    }
                }
            }
            progress(mapOf("stage" to "verifying", "file" to f.name, "received" to f.size, "total" to f.size, "percent" to 100))
            if (!verifySha256(out, f.sha256)) throw IOException("SHA256 mismatch: ${f.name}")
        }
        progress(mapOf("stage" to "done", "file" to "", "received" to 0, "total" to 0, "percent" to 100))
    }

    fun isReady(manifest: Manifest): Boolean {
        if (!rootDir.exists()) return false
        return try {
            manifest.files.all { f ->
                val out = File(rootDir, f.name)
                out.exists() && verifySha256(out, f.sha256)
            }
        } catch (e: Exception) { false }
    }

    private fun copyAsset(path: String, dst: File) {
        dst.parentFile?.mkdirs()
        context.assets.open(path).use { `in` ->
            FileOutputStream(dst).use { out ->
                `in`.copyTo(out)
            }
        }
    }

    private fun verifySha256(file: File, hex: String): Boolean {
        val md = MessageDigest.getInstance("SHA-256")
        DigestInputStream(FileInputStream(file), md).use { it.copyTo(OutputStream.nullOutputStream()) }
        val digest = md.digest().joinToString("") { "%02x".format(it) }
        val expected = hex.trim().lowercase()
        val actual = digest.trim().lowercase()
        return expected == actual
    }

    private fun downloadTo(urlStr: String, dst: File, total: Long, onProg: (rec: Int, tot: Int, pct: Int) -> Unit) {
        dst.parentFile?.mkdirs()
        val tmp = File(dst.absolutePath + ".part")
        var conn: HttpURLConnection? = null
        var received = 0L
        try {
            conn = URL(urlStr).openConnection() as HttpURLConnection
            conn.connectTimeout = 15000; conn.readTimeout = 300000
            conn.inputStream.use { input ->
                FileOutputStream(tmp).use { out ->
                    val buf = ByteArray(8192)
                    while (true) {
                        val n = input.read(buf)
                        if (n <= 0) break
                        out.write(buf, 0, n)
                        received += n
                        val pct = if (total > 0) ((received * 100) / total).toInt() else 0
                        onProg(received.toInt(), total.toInt(), pct)
                    }
                }
            }
            if (dst.exists()) dst.delete()
            tmp.renameTo(dst)
        } finally {
            conn?.disconnect()
            if (tmp.exists() && (!dst.exists() || tmp.length() != dst.length())) tmp.delete()
        }
    }
}

File: learnsynth_offline_llm/ios/Classes/ModelManager.swift
import Foundation
import CommonCrypto

struct ManifestFile: Codable { let name: String; let size: Int; let sha256: String; let url: String?; let asset: Bool?; let demo: Bool? }
struct Manifest: Codable { let ctxLen: Int?; let tokenizer: String?; let files: [ManifestFile] }

class ModelManager {
  let modelId: String
  let rootDir: URL
  init(modelId: String) {
    self.modelId = modelId
    let appSup = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
    self.rootDir = appSup.appendingPathComponent("models/\(modelId)", isDirectory: true)
    try? FileManager.default.createDirectory(at: rootDir, withIntermediateDirectories: true)
  }

  func modelDir() -> URL { rootDir }

  func isReady(_ manifest: Manifest) -> Bool {
    for f in manifest.files {
      let path = rootDir.appendingPathComponent(f.name)
      guard FileManager.default.fileExists(atPath: path.path),
            let ok = try? verifySha256(file: path, hex: f.sha256), ok
      else { return false }
    }
    return true
  }

  func ensureModel(manifest: Manifest, progress: @escaping ([String: Any]) -> Void) throws {
    for f in manifest.files {
      let dst = rootDir.appendingPathComponent(f.name)
      if f.asset == true {
        progress(["stage":"copy-asset","file":f.name,"received":0,"total":f.size,"percent":0])
        try copyAsset(named: "assets/models/\(modelId)/\(f.name)", to: dst)
      } else if (f.demo ?? false) {
        try FileManager.default.createDirectory(at: dst.deletingLastPathComponent(), withIntermediateDirectories: true)
        FileManager.default.createFile(atPath: dst.path, contents: Data())
        progress(["stage":"demo-generate","file":f.name,"received":f.size,"total":f.size,"percent":100])
      } else {
        if !FileManager.default.fileExists(atPath: dst.path) || !(try? verifySha256(file: dst, hex: f.sha256))! {
          guard let urlStr = f.url, let url = URL(string: urlStr) else { throw NSError(domain:"mm", code:1) }
          try download(url: url, to: dst, total: f.size) { rec, tot, pct in
            progress(["stage":"downloading","file":f.name,"received":rec,"total":tot,"percent":pct])
          }
        }
      }
      progress(["stage":"verifying","file":f.name,"received":f.size,"total":f.size,"percent":100])
      if try !verifySha256(file: dst, hex: f.sha256) { throw NSError(domain:"mm", code:2, userInfo:[NSLocalizedDescriptionKey:"SHA mismatch \(f.name)"]) }
    }
    progress(["stage":"done","file":"","received":0,"total":0,"percent":100])
  }

  private func copyAsset(named: String, to dst: URL) throws {
    let bundle = Bundle.main
    guard let src = bundle.url(forResource: named, withExtension: nil) else { throw NSError(domain:"mm", code:3) }
    try FileManager.default.createDirectory(at: dst.deletingLastPathComponent(), withIntermediateDirectories: true)
    if FileManager.default.fileExists(atPath: dst.path) { try FileManager.default.removeItem(at: dst) }
    try FileManager.default.copyItem(at: src, to: dst)
  }

  private func verifySha256(file: URL, hex: String) throws -> Bool {
    let h = try sha256(url: file)
    let expected = hex.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    let actual = h.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    return expected == actual
  }

  private func sha256(url: URL) throws -> String {
    let bufSize = 32768
    var ctx = CC_SHA256_CTX(); CC_SHA256_Init(&ctx)
    let fh = try FileHandle(forReadingFrom: url)
    while autoreleasepool(invoking: {
      let data = fh.readData(ofLength: bufSize)
      if data.count > 0 {
        data.withUnsafeBytes { _ = CC_SHA256_Update(&ctx, $0.baseAddress, CC_LONG(data.count)) }
        return true
      }
      return false
    }) {}
    var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    CC_SHA256_Final(&digest, &ctx)
    return digest.map { String(format: "%02x", $0) }.joined()
  }

  private func download(url: URL, to dst: URL, total: Int, onProg: (Int, Int, Int) -> Void) throws {
    let tmp = dst.appendingPathExtension("part")
    try FileManager.default.createDirectory(at: dst.deletingLastPathComponent(), withIntermediateDirectories: true)
    let data = try Data(contentsOf: url)
    try data.write(to: tmp, options: .atomic)
    onProg(total, total, 100)
    if FileManager.default.fileExists(atPath: dst.path) { try FileManager.default.removeItem(at: dst) }
    try FileManager.default.moveItem(at: tmp, to: dst)
  }
}
